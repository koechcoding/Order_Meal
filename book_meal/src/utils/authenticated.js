export default function authenticated(){
    try{
        return JSON.parse(localStorage.getItem('user'));
    } catch(err){
        
    }
}