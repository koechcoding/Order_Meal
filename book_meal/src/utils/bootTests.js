import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter()});

global.window = {}
import localStorage from 'mock-local-Storage';
window.localStorage = global.localStorage

export default Enzyme;
