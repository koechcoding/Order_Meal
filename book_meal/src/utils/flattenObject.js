export default function flattenObject(obj){
    let result = {}
    for (const key in obj){
        if((typeof obj[key])=== 'object' &&  obj[key]  !== null) {
            let flattened = flattenObject(obj[key])
            for(const key2 in flattened){
                result[key + "." + key2] = flattened[key2];
            }
        }else{
            result[key] = obj[key];
        }
    }
    return result;
}