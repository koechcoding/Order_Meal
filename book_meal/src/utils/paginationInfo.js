export default function paginationInfo(page){

    if(!page){
        return{}
    }

    let currentPage = 1
    if(page.next_page)
         currentPage = page.next_page -1;
    else if (page.prev_page)
        currentPage = page.prev_page + 1;

    let pageCount = Math.ceil(page.total / page.per_paage);
    if(Number.isNaN(pageCount) || pageCount === 0)
      pageCount = 1; 

    return {
        pageCount,
        currentPage,
        total: page.total,
        perPage: page.per_page,
        hasNext: page.has_next,
        hasPrev: page.has_prev,
        nextPage: page.next_page,
        prevPage: page.prev_page,
        currentCount: page.current_count,
    };
}