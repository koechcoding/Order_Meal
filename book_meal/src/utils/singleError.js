const singleError = (errorPayload)=>{
    if(errorPayload && errorPayload.data){
        const response = errorPayload.data;
        //alternative to Object.keys...
        for(let errorKey in response.errors){
            return response.error[errorKey]
        }
        if(response.message){
            return response.message;
        }
    }
    return 'Unknown error. Please try again!';
}
export default singleError;