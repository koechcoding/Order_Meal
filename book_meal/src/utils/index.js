import singleError from './singleError';
import paginationInfo from './paginationInfo';
import authenticated from './authenticated';
import flattenObject from './flattenObject';

export { 
    singleError, 
    paginationInfo, 
    authenticated,
    flattenObject
}