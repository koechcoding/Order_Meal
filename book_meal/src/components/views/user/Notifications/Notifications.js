import React from 'react';
import axios from 'src/axios';
import PropTypes from 'prop-types';
import { Alert } from 'reactstrap';
import Filter from 'src/components/common/Filter';
import Content from 'src/components/common/Content';
import Paginator from 'src/components/common/Paginator';
import ClearModal from './components/Clear';
import { paginationInfo, singleError } from 'src/utils';


class Notifications extends React.Component {

    state = {
        data: {},
        search: '',
        perPage: 6,
    }

    componentWillMount() {
        this.fetchNotifications();
    }

    fetchNotifications = (config = {}) => {
        this.props.setLoading(true);
        const { 
            page = this.state.page,
            search = this.state.search,
            perPage = this.state.perPage,
        } = config;

        const link = 
            `/notifications?per_page=${perPage}&search=${search}&page=${page}`;
        axios.get(link).then(({ data }) => {
            this.setState({
                ...this.state,
                data,
                pageInfo: paginationInfo(data)
            });
            this.props.setLoading(false);
        }).catch(({ response }) =>  {
            this.props.setLoading(false);
        })
    }

    onFilter = (text) => {
        this.setState({
            ...this.state,
            search: text,
        });
        this.fetchNotifications({search: text});
    }

    onPageChange = (page) => {
        this.setState({
            ...this.state,
            page,
        });

        this.fetchNotifications({
            page: page,
            search: this.state.search
        });
    }

    toggleClear = () => {
        this.setState({
            ...this.state,
            clearIsOpen: !this.state.clearIsOpen
        });
    }

    onClear = (id) => {
        axios.auth();
        this.props.setLoading(true);
        let link = null;
        if (id) {
            link = `notifications/${id}`;
        } else {
            link = 'notifications';
        }
        axios.delete(link).then(() => {
            this.fetchNotifications();
            this.props.setLoading(false);
        }).catch(({ response }) => {
            this.setState({
                ...this.state,
                error: response
            });
            this.props.setLoading(false);
        });
    }

    render() {
        const contentTop = (
            <div className="col-12 mb-2 pr-0 pr-sm-2">
                <h5 className="d-inline-block">Notifications</h5>
                <button onClick={this.toggleClear} className="btn btn-secondary float-right">
                    Clear All
                </button>
            </div>
        );

        const contentFilter = (
            <Filter onFilter={this.onFilter} />
        );

        const { clearIsOpen, pageInfo, error } = this.state;
        const { notifications = [] } = this.state.data;
        return (
            <Content 
                {...this.props}
                contentTop={contentTop} 
                contentFilter={contentFilter}>
                {error &&
                    <Alert className="text-center text-small" color="danger">
                        { singleError(error) }
                    </Alert>
                }

                <div className="row">
                    {notifications.map((notice) => (
                        <div key={notice.id} className="col-12 col-sm-6 p-2">
                            <div className="notice p-2">
                                <span onClick={() => this.onClear(notice.id)} className="notice-close">&times;</span>
                                <h5>{notice.title}</h5>
                                <p>{notice.message}</p>
                            </div>
                        </div>
                    ))}

                </div>
                {notifications.length === 0 &&
                    <p className="text-center">You have no notifications for now.</p>
                }

                <ClearModal 
                    {...this.props}
                    onChange={this.fetchNotifications}
                    isOpen={clearIsOpen} 
                    toggle={this.toggleClear}/>

                {notifications.length !== 0 && 
                    <Paginator 
                        onPrev={this.onPageChange}
                        onNext={this.onPageChange} 
                        pageInfo={pageInfo} />
                }
            </Content>
        );
    }
}

Notifications.propTypes = {
    setLoading: PropTypes.func.isRequired,
}

export default Notifications;