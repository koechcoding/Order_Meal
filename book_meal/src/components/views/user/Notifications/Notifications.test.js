import React from 'react';
import { shallow }  from 'enzyme';
import 'src/utils/bootTests';
import Notifications from './Notifications';

const props = {
    toggleMenu: () => {},
    setLoading: () => {}, 
}
describe('<Notifications />', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<Notifications {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });
    it('toggles clear notifications modal', () => {
        const wrapper = shallow(<Notifications {...props}/>);
        const prevState = wrapper.state('clearIsOpen');
        wrapper.instance().toggleClear();
        expect(wrapper.state('clearIsOpen')).toBe(!prevState);
    });
    it('sets current page', () => {
        const wrapper = shallow(<Notifications {...props}/>);
        wrapper.instance().onPageChange(2);
        expect(wrapper.state('page')).toBe(2);
    });
});
