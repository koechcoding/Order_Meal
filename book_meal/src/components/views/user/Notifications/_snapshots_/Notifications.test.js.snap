// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Notifications
    setLoading={[Function]}
    toggleMenu={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        undefined,
        <div
          className="row"
>
          
</div>,
        <p
          className="text-center"
>
          You have no notifications for now.
</p>,
        <ClearModal
          isOpen={undefined}
          onChange={[Function]}
          setLoading={[Function]}
          toggle={[Function]}
          toggleMenu={[Function]}
/>,
        false,
      ],
      "contentFilter": <Filter
        onFilter={[Function]}
/>,
      "contentTop": <div
        className="col-12 mb-2 pr-0 pr-sm-2"
>
        <h5
                className="d-inline-block"
        >
                Notifications
        </h5>
        <button
                className="btn btn-secondary float-right"
                onClick={[Function]}
        >
                Clear All
        </button>
</div>,
      "setLoading": [Function],
      "toggleMenu": [Function],
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "You have no notifications for now.",
          "className": "text-center",
        },
        "ref": null,
        "rendered": "You have no notifications for now.",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "isOpen": undefined,
          "onChange": [Function],
          "setLoading": [Function],
          "toggle": [Function],
          "toggleMenu": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          undefined,
          <div
            className="row"
>
            
</div>,
          <p
            className="text-center"
>
            You have no notifications for now.
</p>,
          <ClearModal
            isOpen={undefined}
            onChange={[Function]}
            setLoading={[Function]}
            toggle={[Function]}
            toggleMenu={[Function]}
/>,
          false,
        ],
        "contentFilter": <Filter
          onFilter={[Function]}
/>,
        "contentTop": <div
          className="col-12 mb-2 pr-0 pr-sm-2"
>
          <h5
                    className="d-inline-block"
          >
                    Notifications
          </h5>
          <button
                    className="btn btn-secondary float-right"
                    onClick={[Function]}
          >
                    Clear All
          </button>
</div>,
        "setLoading": [Function],
        "toggleMenu": [Function],
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "You have no notifications for now.",
            "className": "text-center",
          },
          "ref": null,
          "rendered": "You have no notifications for now.",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "isOpen": undefined,
            "onChange": [Function],
            "setLoading": [Function],
            "toggle": [Function],
            "toggleMenu": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;